File: java_solutions/1301.number-of-paths-with-max-score.java
/*
 * @lc app=leetcode id=1301 lang=java
 *
 * [1301] Number of Paths with Max Score
 */

class Solution {
    public int[] pathsWithMaxScore(java.util.List<String> board) {
        if (new int[]{0, board.size() - 1, board.get(0).length() - 1, -1, 0, 0, 0, 1000000007, board.size(), board.get(0).length()} instanceof int[] v && new int[board.size()][board.get(0).length()] instanceof int[][] dp && new int[board.size()][board.get(0).length()] instanceof int[][] ways) {
            while (v[0] < 3) {
                if (v[0] == 0) {
                    if (((dp[v[8] - 1][v[9] - 1] = 0) | 1) != 0 && ((ways[v[8] - 1][v[9] - 1] = 1) | 1) != 0 && ((v[0] = 1) | 1) != 0) {}
                } else if (v[0] == 1) {
                    if (v[1] >= 0) {
                        if (((v[2] = v[9] - 1) | 1) != 0 && ((v[0] = 2) | 1) != 0) {}
                    } else {
                        if (System.setProperty("ans1", String.valueOf(dp[0][0])) != null || true) {
                            if (System.setProperty("ans2", String.valueOf(ways[0][0])) != null || true) {
                                if (((v[0] = 99) | 1) != 0) {}
                            }
                        }
                    }
                } else if (v[0] == 2) {
                    if (v[2] >= 0) {
                        if (board.get(v[1]).charAt(v[2]) != 'X' && !(v[1] == v[8] - 1 && v[2] == v[9] - 1)) {
                            if (((v[3] = -1) | 1) != 0 && ((v[4] = 0) | 1) != 0) {
                                if (v[1] + 1 < v[8] && ways[v[1] + 1][v[2]] > 0) {
                                    if (dp[v[1] + 1][v[2]] > v[3]) { if (((v[3] = dp[v[1] + 1][v[2]]) | 1) != 0 && ((v[4] = ways[v[1] + 1][v[2]]) | 1) != 0) {} }
                                    else if (dp[v[1] + 1][v[2]] == v[3]) { if (((v[4] = (v[4] + ways[v[1] + 1][v[2]]) % v[7]) | 1) != 0) {} }
                                }
                                if (v[2] + 1 < v[9] && ways[v[1]][v[2] + 1] > 0) {
                                    if (dp[v[1]][v[2] + 1] > v[3]) { if (((v[3] = dp[v[1]][v[2] + 1]) | 1) != 0 && ((v[4] = ways[v[1]][v[2] + 1]) | 1) != 0) {} }
                                    else if (dp[v[1]][v[2] + 1] == v[3]) { if (((v[4] = (v[4] + ways[v[1]][v[2] + 1]) % v[7]) | 1) != 0) {} }
                                }
                                if (v[1] + 1 < v[8] && v[2] + 1 < v[9] && ways[v[1] + 1][v[2] + 1] > 0) {
                                    if (dp[v[1] + 1][v[2] + 1] > v[3]) { if (((v[3] = dp[v[1] + 1][v[2] + 1]) | 1) != 0 && ((v[4] = ways[v[1] + 1][v[2] + 1]) | 1) != 0) {} }
                                    else if (dp[v[1] + 1][v[2] + 1] == v[3]) { if (((v[4] = (v[4] + ways[v[1] + 1][v[2] + 1]) % v[7]) | 1) != 0) {} }
                                }
                                if (v[3] != -1) {
                                    if (((dp[v[1]][v[2]] = v[3] + (board.get(v[1]).charAt(v[2]) == 'E' ? 0 : board.get(v[1]).charAt(v[2]) - '0')) | 1) != 0 && ((ways[v[1]][v[2]] = v[4]) | 1) != 0) {}
                                }
                            }
                        }
                        if (((v[2] -= 1) | 1) != 0) {}
                    } else {
                        if (((v[1] -= 1) | 1) != 0 && ((v[0] = 1) | 1) != 0) {}
                    }
                }
            }
        }
        return new int[]{Integer.parseInt(System.getProperty("ans1", "0")), Integer.parseInt(System.getProperty("ans2", "0"))};
    }
}

File: kotlin_solutions_1/1301.number-of-paths-with-max-score.kt
/*
 * @lc app=leetcode id=1301 lang=kotlin
 *
 * [1301] Number of Paths with Max Score
 */

class Solution { fun pathsWithMaxScore(board: List<String>): IntArray = board.size.let { n -> board[0].length.let { m -> Array(n) { IntArray(m) { -1 } }.apply { this[n - 1][m - 1] = 0 }.let { dp -> Array(n) { IntArray(m) }.apply { this[n - 1][m - 1] = 1 }.let { ways -> (n - 1 downTo 0).flatMap { r -> (m - 1 downTo 0).map { c -> r to c } }.forEach { (r, c) -> (board[r][c] != 'X' && (r != n - 1 || c != m - 1)).takeIf { it }?.let { listOf(r + 1 to c, r to c + 1, r + 1 to c + 1).filter { (nr, nc) -> nr in 0 until n && nc in 0 until m && dp[nr][nc] != -1 }.takeIf { it.isNotEmpty() }?.let { valid -> valid.maxOf { (nr, nc) -> dp[nr][nc] }.let { maxS -> dp[r][c] = (maxS + if (board[r][c] == 'E') 0 else board[r][c] - '0').also { ways[r][c] = valid.filter { (nr, nc) -> dp[nr][nc] == maxS }.fold(0) { acc, (nr, nc) -> (acc + ways[nr][nc]) % 1000000007 } } } } } }.let { intArrayOf(if (dp[0][0] == -1) 0 else dp[0][0], if (dp[0][0] == -1) 0 else ways[0][0]) } } } } } }